File: src/main/java/com/zunza/buythedip/watchlist/repository/WatchlistRepository.java
package com.zunza.buythedip.watchlist.repository;

import java.util.List;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import com.zunza.buythedip.watchlist.dto.WatchlistDetailsResponse;
import com.zunza.buythedip.watchlist.dto.WatchlistResponse;
import com.zunza.buythedip.watchlist.entity.Watchlist;

@Repository
public interface WatchlistRepository extends JpaRepository<Watchlist, Long> {
	@Query(
		"""
		SELECT new com.zunza.buythedip.watchlist.dto.WatchlistResponse(
		w.id,
		w.name,
		w.isDefault,
		w.isSystem,
		w.sortOrder
		)
		From Watchlist w
		WHERE (:userId IS NULL AND w.isSystem = true)
		OR (:userId IS NOT NULL AND w.user.id = :userId)
		ORDER By w.sortOrder ASC
		"""
	)
	List<WatchlistResponse> findWatchlistsByUserId(@Param("userId") Long userId);

	@Query(
		"""
		SELECT new com.zunza.buythedip.watchlist.dto.WatchlistDetailsResponse(
		wi.id,
		c.id,
		c.name,
		c.symbol,
		m.logo,
		wi.sortOrder
		)
		FROM Watchlist w
		JOIN w.watchlistItems wi
		JOIN wi.crypto c
		JOIN c.metadata m
		WHERE w.id = :watchlistId
		ORDER BY wi.sortOrder ASC
		"""
	)
	List<WatchlistDetailsResponse> findWatchlistDetailsById(@Param("watchlistId") Long watchlistId);
}

File: src/main/kotlin/com/zunza/buythedip/watchlist/repository/WatchlistRepository.kt
package com.zunza.buythedip.watchlist.repository

import com.zunza.buythedip.watchlist.dto.WatchlistDetailsResponse
import com.zunza.buythedip.watchlist.entity.Watchlist
import com.zunza.buythedip.watchlist.repository.querydsl.WatchlistQuerydslRepository
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository

@Repository
interface WatchlistRepository : JpaRepository<Watchlist, Long>, WatchlistQuerydslRepository {

    @Query(
        """
            SELECT new com.zunza.buythedip.watchlist.dto.WatchlistDetailsResponse(
            c.id,
            c.name,
            c.symbol,
            c.logo,
            wi.sortOrder
            )
            FROM Watchlist w
            JOIN w.watchlistItems wi
            JOIN wi.crypto c
            WHERE w.id = :watchlistId
            ORDER BY wi.sortOrder ASC
        """
    )
    fun findWatchlistDetailsById(@Param("watchlistId") watchlistId: Long): List<WatchlistDetailsResponse>
}
